language/libra-vm/src/libra_vm.rs

```diff
@@ -442,9 +442,9 @@ pub fn txn_effects_to_writeset_and_events_cached<C: AccessPathCache>(
             let op = match val_opt {
                 None => WriteOp::Deletion,
                 Some((ty_layout, val)) => {
-                    let blob = val
-                        .simple_serialize(&ty_layout)
-                        .ok_or_else(|| VMStatus::Error(StatusCode::VALUE_SERIALIZATION_ERROR))?;
+                    let blob = val.simple_serialize(&ty_layout).ok_or_else(|| {
+                        VMStatus::Error(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
```

Comment:
The Move VM should never return a layout that doesn't match the value.

Reply:
out of curiosity, isn't the case that both of those generate a discarded transaction with an invariant violation?
I guess what I am asking is "would this change makes it easier or harder to understand what is going on?",
or maybe you are saying that this being an impossible condition it may be better to return a generic invariant error?

Reply:
None of `VALUE_SERIALIZATION_ERROR`, `VALUE_DESERIALIZATION_ERROR` and `DATA_FORMAT_ERROR` is in the invariant violation category and they will incorrectly label transaction outputs as Kept.
```
                match code.status_type() {
                    // Any unknown error should be discarded
                    StatusType::Unknown => Err(code),
                    // Any error that is a validation status (i.e. an error arising from the prologue)
                    // causes the transaction to not be included.
                    StatusType::Validation => Err(code),
                    // If the VM encountered an invalid internal state, we should discard the transaction.
                    StatusType::InvariantViolation => Err(code),
                    // A transaction that publishes code that cannot be verified will be charged.
                    StatusType::Verification => Ok(KeptVMStatus::MiscellaneousError),
                    // If we are able to decode the`SignedTransaction`, but failed to decode
                    // `SingedTransaction.raw_transaction.payload` (i.e., the transaction script),
                    // there should be a charge made to that user's account for the gas fees related
                    // to decoding, running the prologue etc.
                    StatusType::Deserialization => Ok(KeptVMStatus::MiscellaneousError),
                    // Any error encountered during the execution of the transaction will charge gas.
                    StatusType::Execution => Ok(KeptVMStatus::ExecutionFailure {
                        location: AbortLocation::Script,
                        function: 0,
                        code_offset: 0,
                    }),
                }
```

Reply:
got it, fantastic!